midi/midi/encode.py

```diff
@@ -4,227 +4,245 @@
+        for i, value in enumerate(event):
+            if value:  # True or > 0.0
+                notes.append((i, min(127, round(value * 128))))
+        events.append(notes)
+
+    return events
 
-    if input_array.shape[1] + 1 != len(tempos):
-        raise IndexError('length of input array\'s time dimension must be shorter by 1 than tempos array')
 
+def get_messages_from_tuples(track, track_channel, event_lengths,
+                             accuracy, join_notes, use_default_velocity, default_velocity=DEFAULT_VELOCITY):
+    """
+    translates a list of tuples of active notes into a MidiTrack
+    """
+    if len(track) != len(event_lengths):
+        raise IndexError('input event length array and data event dimension must be of equal length')
+
+    messages = []
+    last_event = []
+    time = 0
+    for i, current_event in enumerate(track):
+        if join_notes:
+            subtract = [elem for elem in last_event if elem not in current_event]
+            add = [elem for elem in current_event if elem not in last_event]
+
+            for elem in subtract:
+                messages.append(Message('note_off', note=elem[0], channel=track_channel,
+                                        velocity=0, time=round(time * accuracy)))
+                time = 0
+
+            for elem in add:
+                messages.append(Message('note_on', note=elem[0], channel=track_channel,
+                                        velocity=default_velocity if use_default_velocity else elem[1],
+                                        time=round(time * accuracy)))
+                time = 0
+
+        else:
+            for elem in last_event:
+                messages.append(Message('note_off', note=elem[0], channel=track_channel,
+                                        velocity=0, time=round(time * accuracy)))
+                time = 0
+
+            for elem in current_event:
+                messages.append(Message('note_on', note=elem[0], channel=track_channel,
+                                        velocity=default_velocity if use_default_velocity else elem[1],
+                                        time=round(time * accuracy)))
+                time = 0
+
+        last_event = current_event
+        time += event_lengths[i]
+
+    for elem in last_event:
+        messages.append(Message('note_off', note=elem[0], channel=track_channel,
+                                velocity=0, time=round(time * accuracy)))
+        time = 0
+
+    messages.append(MetaMessage('end_of_track', time=0))
+    return MidiTrack(messages)
+
+
+def prepare_meta_file(tempos, grid_accuracy, event_lengths=None):
     midi_file = MidiFile(ticks_per_beat=240)
-    acc_factor = float(960 / accuracy)  # acc_factor = ticks_per_beat / (notated_32nd_notes_per_beat * accuracy / 32)
-    midi_file.tracks.append(get_tempo_meta_messages(tempos, acc_factor))
+    accuracy = float(960 / grid_accuracy)  # = ticks_per_beat / (notated_32nd_notes_per_beat * grid_accuracy / 32)
```

Comment:
Please remind me: `ticks_per_beat=240` is that something we can hardcode? Is it dependent on some variable? Also, what is this `/ 32` number?

Reply:
`ticks_per_beat=240` is a choice of mine, `32` is notated 32nd notes per measure

Reply:
Since this `960` that was calculated depending on `240`, please extract `240` into some constant (does not have to be global) and calculate accuracy in full step

Reply:
What info gain will you have from parameterization of 240? It is already described by the name of an argument it is assigned to, 960 is explained in the comment. This is optimal both time- and memory-wise.

Reply:
changed accuracy calculation to descriptive form in ed79d5dcd403e9b7edea21c6781f58840aead36f